### ISSUE TITLE
Notes of optics.py

### ISSUE THREAD

**Issue by ZERO2ER0** [ZERO2ER0/opticalCNN#2]
# optics.py is the corn codes of the optical convolution neural network.

## Description 

optics.py 中的函数有基本的*傅里叶变换函数*、*光学元件（optics elements）函数*、*光学传播函数*。其中最重要是*光学传播函数*。

## Plan

对optics.py文件中的函数进行学习和注释。包括但不限于对应的光学公式，以及应用范围。

## Helper functions

- `zoom(image_batch, zoom_fraction)`
  Description: Get central crop of batch using `tf.images.central_crop(image, central_fraction)`

- `transp_fft2d(a_tensor, dtype=tf.complex128)`
  Description: Takes images of shape [batch_size, x, y, channels] and transposes them correctly for tensorflows fft2d to work, Computes the 2-dimensional discrete Fourier transform over the inner-most 2 dimensions of input

- `transp_ifft2d(a_tensor, dtype=tf.complex128)`
  Description: 二维逆向傅里叶变换，其中 `tf.cast`类型转换，`tf.transpose` 维数置换

- `compl_exp_tf(phase, dtype=tf.complex128, name='complex_exp')`
  Description: 将phase转换为复指数形式

- `laplacian_filter_tf(img_batch)`
  Description: 拉普拉斯滤波器。

  Function：边缘提取。

- `laplace_l1_regularizer(scale)`
  Description: laplacian L1 regularizer, $J(\theta , X, y)= L_{emp} (\theta ; X, y) + \alpha \Omega (\theta)$，return $\alpha \Omega(\theta)$，其中$\Omega(\theta)=tf.reduce\_mean$。

  Function: 降维

- `phaseshifts_from_height_map(height_map, wave_lengths, refractive_index)`

  Description: Calculates the phase shifts created by a height map with certain refractive index for light with specifice wave length. 计算由 height map 引起的 phase shifts。

  Optics: $\Delta N = n -1.000277, k = \frac{2\pi}{\lambda}, \phi = k\times \Delta N \times height\_map$，其中 $\Delta N \times height\_map $是光程， return 复指数形式的phase shifts。

- `get_one_phase_shift_thickness(wave_lengths, refractive_index)`

  Description: Calculate the thickness (in meter) of a phaseshift of 2pi. 计算光在介质中传播一个周期也就是$2\pi$相移的厚度

  Optics: $\Delta N = n -1, k = \frac{2\pi}{\lambda}, thickness\ of\  2\pi = \frac{2\pi}{k \Delta N}\times 1.1$，与`pahseshifts_from_height_map`互为反函数。

- `attach_summaries(name, var, image=False, log_image=False)`

  Description: Attaching a tensor summary will allow us to retrieve the actual value of the height map just from the summary

- `attach_img(name, var)`

- `fftshift2d_tf(a_tensor)`

  Description: 将零频分量平移到频谱中心

- `ifftshift2d_tf(a_tensor)`

  Description: 逆零频平移，撤销`fftshift2d_tf()`的结果。

- `psf2otf(input_filter, output_size)`

  Description: Convert 4D tensor flow filter into its FFT.  对filter进行circularly shift，在进行傅里叶变换，得到filter的FFT变换。

  Function: Circularly shift so center pixel is at 0,0.

  - [ ] Question: why the name of this function is `psf2otf`?

- `fft_conv2d(img, psf, adjoint=False)`

  Description: Implements convolution in the frequency domain, with circular boundary conditions. 在圆形边界条件下的频域中实施卷积。

  Optics: 

  ​	圆形边界条件具有low pass filter特性。

  1. img零频分量平移
  2. img的fft变换
  3. otf与img频域相乘
  4. ifft
  5. 逆零频平移

- `depth_dep_convolution(img, psfs, disc_depth_map, use_fft=True)`

  Description: Pick each pixel from the blur volume according to the disc_depth_map. 根据离散的深度图，从在频域中卷积完的blurred images中提取像素。

  - [ ] Question: 深度图有什么作用？为什么要根据他的索引来提取blurred images像素。

- `get_spherical_wavefront_phase(resolution,
      physical_size,
      wave_lengths,
      source_distance)`

  Description: 球面波的波前相位。

  Optics:

  ​	$r = \sqrt{x^2+y^2+z^2},\ k=\frac{2\pi}{\lambda}$，其中$z$为source distance, r 是矢径，在这里即光程。

  ​	$\phi=k\times r$ 相位 = 光程x波矢

  ​	$\phi$  复指数形式

- `least_common_multiple(a, b)`

  Descriptiomn: least common multiple, 最小公倍数。

- `area_downsampling_tf(input_image, target_side_length):`

  Description: downsampling, 下采样。

- `get_intensities(input_field)`

  Description: get intensities,  得到强度，比如光强。

- `gaussian_noise(image, stddev=0.001)`

  Description: return the images with gaussian noise.

## Optical elements & Propagation

- `class Propagation(abs.ABC)`

  - `__init__(self, input_shape, distance, discretization_size, wave_lengths, pad_width, pad_mode)`
  - `-propagate(self, input_filed)`

  Description: 

  1. import abc 利用abc实现抽象类。
  2. @abc.abstractemethod 定义抽象方法，无需实现功能。

- `class FresnelPropagation(Propagation)

  - `_propagate(self, input_field)`

  Description: 

  1. 继承父类Propagation。

  Functions:

  1. Zero padding.
  2. $f_x=\frac{x}{d\times N}$,  其中d是discretization size 输入的采样间隔（input sample interval），N是输入域的size，$f_x，f_y$为与空间坐标对应的品与坐标，个人理解可以看成comb函数，二维抽样函数。  信息光学教程中$f_x=\frac{x}{\lambda f}, f_y=\frac{y}{\lambda f}$，其中f为透镜焦距，显然与程序中不同。
  3. 最大的空间频率为$max\ frequency = \frac{1}{2\times pixel\_size}$.
  4. $f_x= ifftshift(f_x), f_y= ifftshift(f_y)$
  5. $tmp = -\lambda \pi (f_x^2+f_y^2)  = -\frac{k\lambda ^2}{2}(f_x^2+f_y^2)  $, Fresnel kernel constant exponent part.
  6. $d\times tmp=   -kd\frac{\lambda ^2}{2}(f_x^2+f_y^2)$
  7. $H=cos(d\times tmp)+jsin(d\times tmp)$ , Fresnel kernel. 
  8. $objFT = fft(padded\_input\_field$)
  9. $out\_filed = ifft(objFT* H)$

  Opitcs:

  Fresnel 衍射传递函数
  $$
  H_F(f_x,f_y)=exp[jkd(1-\frac{\lambda ^2}{2}(f_x^2+f_y^2))
  $$
  其中k是波矢的标量，即光波数，d是衍射距离.

  衍射的菲涅尔近似表示为
  $$
  U(x,y,d) = F^{-1}\{F\{U(x,y,0)\}H_F(f_x,f_y)\}
  $$
  

- `class AmplitudeMask()`

  - `__call__(self, input_field)`

    Descripition:

    返回dtype = complex128

    $Amp\_mask= input\_filed \times amp\_mask$

- `bacteriorhodopsin_nonlinearity(input_field, trainable=False)`

  Description:

  Bacteriorhodopsin_nonlinearity: 细菌视紫红外质 是一种蛋白质，他能捕获光能并将质子移出细胞外，随后将得到的质子梯度转化为化学能。

  Function: 

  $ampltude = |input\_filed|$,

  $ phase = \arctan \frac{Im(input\_filed)}{Re({input\_filed})}, amplitude\_nl = a + b \log{amplitude}$,其中$a = 0.8178, b= 0.03047$

- `class PhasePlate()`

  ​	相位掩膜板

  - `__init__(self,
        wave_lengths,
        height_map,
        refractive_index,
        phase_shifts=None,
        height_tolerance=None,
        lateral_tolerance=None)`

  - `_build(self)`

    Description: phase shift from height map

  - `__call__(self, input_field)`

    Description: 返回 $input \circ phase\_shift$，输入经过掩膜板后的输出。

- `propagate_fresnel(input_field,
      distance,
      input_sample_interval,
      wave_lengths,
      pad_width=None,
      pad_mode=None)`

  Deccription: 菲涅尔衍射传播

- `square_aperture(input_field, aperture_fraction)`

  Description: 方形光栅

- `circular_aperture(input_field, max_val=None)`

  Description: 圆孔光栅

- `phase_shift_element_old(input_field,
      name,
      block_size=1,
      phase_shift_initializer=None,
      phase_shift_regularizer=None,
      refractive_index=1.5)`

  Description: 与phase shift element实现的功能基本一致，未加入高度误差。

- `phase_shift_element(map_shape,
      name,
      wave_lengths,
      block_size=1,
      phase_shift_initializer=None,
      phase_shift_regularizer=None,
      height_tolerance=None, # Default height tolerance is 2 nm.
      refractive_index=1.5)`

  Description: 返回输入图片与复指数形式的相位平移的相乘（逐元素相乘），其中掩膜板加入了高度误差。

  $phase\_shift = compl\_exp\_tf(phase shift) $

- `height_map_element(map_shape,
      name,
      wave_lengths,
      block_size=1,
      height_map_initializer=None,
      height_map_regularizer=None,
      height_tolerance=None, # Default height tolerance is 2 nm.
      refractive_index=1.5)`

  Description: $phase\ shifts= phase\ shift\ from\ height\ map$

### Fourier coefficient phase mask

- `fourier_element(map_shape,
      name,
      wave_lengths,
      refractive_index,
      frequency_range = 0.5,
      height_map_regularizer=None,
      height_tolerance=None, # Default height tolerance is 2 nm.
     )`

  Description: 

  1. 初始化 fourier coefficient real, fourier coefficient complex,

  2. $fourier\ coefficient = real + j\ cplx $

  3. Padding
  4. from fourier_coeffs_padded to height_map, $height\_map = real(ifft(ishift(fourier\ coefficient\ padded)))$
  5. $element = PhasePlate(height\ map)$

### Zernike basis phase mask object

- `zernike_element(zernike_volume,
      name,
      wavelengths,
      refractive_idcs,
      r_NA,
      zernike_initializer=None,
      height_map_regularizer=None,
      height_tolerance=None, # Default height tolerance is 2 nm.
      zernike_scale=1e5)`

  Description: 

  

### Amplitude Mask Object

- `class AmplitudeMask()`

  - `__init__(self)`
  - `_build(self)`
  - `__call__(self, input_field)`

- `amplitude_map_element(map_shape, r_NA,
      name,
      block_size=1,
      amplitude_map_initializer=None,
      amplitude_map_regularizer=None)`

  Description:

  1. 圆形光栅，幅度平方

- `plano_convex_lens(input_field,
      focal_length,
      wave_lengths,
      discretization_step,
      refractive_index=1.5)`

  Description: 

  $convex\ radius = (refractive-1)*f$

  $height map = -\frac{x^2 +y^2}{2}\frac{1}{convex\ radius}$

- `plano_convex_initializer(focal_length,
      wave_lengths,
      discretization_step,
      refractive_index=1.5)`

  Description: 

- `cubic_phase_shifts(input_field)`

  Description: 立方相移

- `propagate_2f_collimator(input_field)`

  Description: transpose fft2d

- `class Sensor()`

  - `__init__(self, resolution, input_is_intensities=False)`
  - `__call__(self, input_field)`

  Description: 

  Sensor readings = area downsampling

- `gaussian_noise(image, stddev=0.001)`

- `class OpticalSystem(abc.ABC)`

  - `__init__(self,
        forward_model,
        wave_resolution,
        wave_lengths,
        sensor_resolution,
        noise_model=gaussian_noise,
        psf_resolution=None,
        target_distance=None,
        discretization_size=1e-6,
        use_planar_incidence=True,
        upsample=True,
        depth_bins=None)`

    Description: 

    1. get psfs
    2. $pixel\ size = Size_{discretization}\times Resolution\_{wave}/ Resolution_{sensor}$，其中$Size_{discretization}$可以看做$input\ sample\ interval$

  - `get_psfs(self)`

    Description:

    1. $size_{physical} = Nsize_{discretization} $，其中N是wave_resolution[0]。
    2. $psf = get\ intensities(sensor\ incident\ field)$，psf 传感器输入的强度

  - `get_sensor_img(self,
        input_img,
        noise_sigma=0.001,
        depth_dependent=False,
        depth_map=None)`

    Description:

    1.  Upsample input image to match wave resolution.

       上采样输入图片来匹配波分辨率

    2. Downsample measured image to match sensor resolution.

       下采样测量图片来匹配传感器分辨率
